frame/mosaic/src/lib.rs

```diff
@@ -0,0 +1,676 @@
+				account.clone(),
+				asset_id,
+				|prev| {
+					let (balance, _) = prev.as_mut().ok_or(Error::<T>::NoClaimableTx)?;
+					// Wipe the entire incoming transaction.
+					if *balance == untrusted_amount {
+						*prev = None;
+					} else {
+						*balance = balance.saturating_sub(untrusted_amount);
+					}
+					T::Assets::burn_from(
+						asset_id,
+						&Self::sub_account_id(SubAccount::incoming(account.clone())),
+						untrusted_amount,
+					)?;
+					Self::deposit_event(Event::<T>::TransferIntoRescined {
+						account,
+						amount: untrusted_amount,
+						asset_id,
+					});
+					Ok(())
+				},
+			)?;
+
+			Ok(().into())
+		}
+
+		/// Collects funds deposited by the relayer into the owner's account
+		#[pallet::weight(10_000)]
+		pub fn claim_to(
+			origin: OriginFor<T>,
+			asset_id: AssetIdOf<T>,
+			to: AccountIdOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let caller = ensure_signed(origin)?;
+			let now = <frame_system::Pallet<T>>::block_number();
+
+			IncomingTransactions::<T>::try_mutate_exists::<_, _, _, DispatchError, _>(
+				caller.clone(),
+				asset_id,
+				|deposit| {
+					let (amount, unlock_after) = deposit.ok_or(Error::<T>::NoClaimableTx)?;
+					ensure!(unlock_after < now, Error::<T>::TxStillLocked);
+					T::Assets::transfer(
+						asset_id,
+						&Self::sub_account_id(SubAccount::incoming(caller.clone())),
+						&to,
+						amount,
+						false,
+					)?;
+					// Delete the deposit.
+					deposit.take();
+					Self::deposit_event(Event::<T>::TransferClaimed {
+						by: caller,
+						to,
+						asset_id,
+						amount,
+					});
+					Ok(())
+				},
+			)?;
+			Ok(().into())
+		}
+	}
+
+	fn ensure_relayer<T: Config>(
+		origin: OriginFor<T>,
+	) -> Result<(RelayerConfig<AccountIdOf<T>, BlockNumberOf<T>>, BlockNumberOf<T>), DispatchError>
+	{
+		let acc = ensure_signed(origin).map_err(|_| DispatchError::BadOrigin)?;
+		let current_block = <frame_system::Pallet<T>>::block_number();
+		let relayer = Relayer::<T>::get().update(current_block);
+		ensure!(relayer.is_relayer(&acc), DispatchError::BadOrigin);
+		Ok((relayer, current_block))
+	}
+
+	#[pallet::extra_constants]
+	impl<T: Config> Pallet<T> {
+		pub fn timelock_period() -> BlockNumberOf<T> {
+			TimeLockPeriod::<T>::get()
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// AccountId of the pallet, used to store all funds before actually moving them.
+		pub fn sub_account_id(sub_account: SubAccount<T>) -> AccountIdOf<T> {
+			T::PalletId::get().into_sub_account(sub_account.to_id())
+		}
+
+		/// Queries storage, returning the account_id of the current relayer.
+		pub fn relayer_account_id() -> Option<AccountIdOf<T>> {
+			let current_block = <frame_system::Pallet<T>>::block_number();
+			Relayer::<T>::get().update(current_block).account_id().cloned()
+		}
+	}
+
+	/// Convenience identifiers emitted by the pallet for relayer bookkeeping.
+	pub type Id = H256;
+
+	/// Raw ethereum addresses.
+	pub type EthereumAddress = [u8; 20];
```

Comment:
we have pallet who has that may be. could share typedef

Reply:
CU-1wanc6n